ddtrace/tracer/stats.go

```diff
@@ -0,0 +1,362 @@
+func (c *concentrator) Stop() {
+	if atomic.SwapUint64(&c.stopped, 1) > 0 {
+		return
+	}
+	close(c.stop)
+	c.wg.Wait()
+}
+
+func (c *concentrator) flush(timenow time.Time) statsPayload {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := timenow.UnixNano()
+	sp := statsPayload{
+		Hostname: c.cfg.hostname,
+		Env:      c.cfg.env,
+		Version:  c.cfg.version,
+		Stats:    make([]statsBucket, 0, len(c.buckets)),
+	}
+	for ts, srb := range c.buckets {
+		if ts > now-int64(c.bufferLen)*c.bucketSize {
+			// Always make sure that we keep the most recent bufferLen number of buckets.
+			// This is a trade-off: we accept slightly late traces (clock skew and stuff)
+			// but we delay flushing by at most bufferLen buckets.
+			continue
+		}
+		log.Debug("Flushing bucket %d", ts)
+		sp.Stats = append(sp.Stats, srb.Export())
+		delete(c.buckets, ts)
+	}
+	// After flushing, update the oldest timestamp allowed to prevent having stats for
+	// an already-flushed bucket.
+	newOldestTs := alignTs(now, c.bucketSize) - int64(c.bufferLen-1)*c.bucketSize
+	if newOldestTs > c.oldestTs {
+		log.Debug("Update oldestTs to %d", newOldestTs)
+		c.oldestTs = newOldestTs
+	}
+	return sp
+}
+
+// aggregation specifies a uniquely identifiable key under which a certain set
+// of stats are grouped inside a bucket.
+type aggregation struct {
+	Name       string
+	Env        string
+	Type       string
+	Resource   string
+	Service    string
+	Hostname   string
+	StatusCode uint32
+	Version    string
+	Synthetics bool
+}
+
+type rawBucket struct {
+	start, duration uint64
+	data            map[aggregation]*rawGroupedStats
+}
+
+func newRawBucket(btime uint64, bsize int64) *rawBucket {
+	return &rawBucket{
+		start:    btime,
+		duration: uint64(bsize),
+		data:     make(map[aggregation]*rawGroupedStats),
+	}
+}
+
+func (sb *rawBucket) handleSpan(ss *aggregableSpan) {
+	gs, ok := sb.data[ss.key]
+	if !ok {
+		gs = newRawGroupedStats()
+		sb.data[ss.key] = gs
+	}
+	if ss.TopLevel {
+		gs.topLevelHits++
+	}
+	gs.hits++
+	if ss.Error != 0 {
+		gs.errors++
+	}
+	gs.duration += float64(ss.Duration)
+	// alter resolution of duration distro
+	trundur := nsTimestampToFloat(ss.Duration)
+	if ss.Error != 0 {
+		gs.errDistribution.Add(trundur)
+	} else {
+		gs.okDistribution.Add(trundur)
+	}
+}
+
+// Export transforms a RawBucket into a statsBucket, typically used
+// before communicating data to the API, as RawBucket is the internal
+// type while statsBucket is the public, shared one.
+func (sb *rawBucket) Export() statsBucket {
+	csb := statsBucket{
+		Start:    sb.start,
+		Duration: sb.duration,
+		Stats:    make([]groupedStats, len(sb.data)),
+	}
+	for k, v := range sb.data {
+		b, err := v.export(k)
+		if err != nil {
+			log.Error("Could not export stats bucket: %v.", err)
+			continue
+		}
+		csb.Stats = append(csb.Stats, b)
+	}
+	return csb
+}
+
+type rawGroupedStats struct {
+	// using float64 here to avoid the accumulation of rounding issues.
```

Comment:
no rounding should be involved here as all values are integers

Reply:
You're right. Changed that.